service-base/src/main/java/org/eclipse/hono/service/credentials/CredentialsEndpoint.java

```diff
@@ -0,0 +1,173 @@
+                    .open();
+        }
+    }
+
+    @Override
+    public void onLinkAttach(final ProtonSender sender, final ResourceIdentifier targetResource) {
+        if (targetResource.getResourceId() == null) {
+            logger.debug("link target provided in client's link ATTACH must not be null, but must match pattern \"credentials/<tenant>/<reply-address>\" instead");
+            sender.setCondition(condition(AmqpError.INVALID_FIELD.toString(),
+                    "link target must not be null but must have the following format credentials/<tenant>/<reply-address>"));
+            sender.close();
+        } else {
+            logger.debug("establishing sender link with client [{}]", MessageHelper.getLinkName(sender));
+            final MessageConsumer<JsonObject> replyConsumer = vertx.eventBus().consumer(targetResource.toString(), message -> {
+                // TODO check for correct session here...?
+                logger.trace("forwarding reply to client: {}", message.body());
+                final Message amqpReply = CredentialsConstants.getAmqpReply(message);
+                sender.send(amqpReply);
+            });
+
+            sender.closeHandler(senderClosed -> {
+                replyConsumer.unregister();
+                senderClosed.result().close();
+                final String linkName = MessageHelper.getLinkName(sender);
+                logger.debug("receiver closed link [{}], removing associated event bus consumer [{}]", linkName, replyConsumer.address());
+            });
+
+            sender.setQoS(ProtonQoS.AT_LEAST_ONCE).open();
+        }
+    }
+
+
+    private void processRequest(final Message msg) {
+
+        final JsonObject credentialsMsg = CredentialsConstants.getCredentialsMsg(msg);
+
+        vertx.eventBus().send(EVENT_BUS_ADDRESS_CREDENTIALS_IN, credentialsMsg,
+                result -> {
+                    JsonObject response = null;
+                    if (result.succeeded()) {
+                        // TODO check for correct session here...?
+                        response = (JsonObject) result.result().body();
+                    } else {
+                        logger.debug("failed to process credentials request [msg ID: {}] due to {}", msg.getMessageId(), result.cause());
+                        // we need to inform client about failure
+                        response = CredentialsConstants.getReply(
+                                HttpURLConnection.HTTP_INTERNAL_ERROR,
+                                MessageHelper.getTenantIdAnnotation(msg),
+                                null,
+                                null);
+                    }
+                    addHeadersToResponse(msg, response);
+                    vertx.eventBus().send(msg.getReplyTo(), response);
+                });
+    }
+
+    private void addHeadersToResponse(final Message request, final JsonObject message) {
```

Comment:
Yes, for sure - just postponed this to the refactoring afterwards. But since we have a base class here already available it is probably better to change it already. Will move it upwards...